src/main/java/com/iggroup/universityworkshopmw/domain/services/OpenPositionsService.java

```diff
@@ -35,46 +35,77 @@ public OpenPositionsService(ClientService clientService) {
       throw new NoAvailableDataException("No open positions exist for client: " + clientId);
    }
 
+   //Keep order: 1)Check funds 2)Add OpenPosition 3)Update client funds
    public OpenPosition addOpenPositionForClient(String clientId, OpenPosition newOpenPosition) throws NoAvailableDataException, InsufficientFundsException {
-      double positionPrice = newOpenPosition.getBuySize() * newOpenPosition.getOpeningPrice();
+      double positionPrice = getPositionOpeningPrice(newOpenPosition);
       List<OpenPosition> openPositionsForClient = clientPositionStore.get(clientId);
 
-      double clientFunds = checkClientFunds(clientId, positionPrice);
+      double clientAvailableFunds = checkClientAvailableFunds(clientId, positionPrice);
       OpenPosition openPositionWithId = updateStoreWithNewPosition(clientId, newOpenPosition, openPositionsForClient);
 
-      clientService.updateFunds(clientId, clientFunds);
+      double newAvailableFunds = calculateNewAvailableFunds(clientAvailableFunds, positionPrice);
+      clientService.updateAvailableFunds(clientId, newAvailableFunds);
```

Comment:
Could we make `updateClientFunds` more generic, so that it can be used here as well (instead of `clientService.updateAvailableFunds`)? Also, if we’re doing that, then it might be better if we have the `clientService.getClientData(clientId);` call separately (to avoid doing it twice) and then passing the client in.

Maybe `updateClientFunds` could have the signature `private void updateClientFunds(Client client, double updatedFunds)`?
And client and updated funds could be calculated outside of this. Or just not have this method and at all and call the update on the service directly.

Reply:
Let me know if that makes no sense

Reply:
We can go through this on Friday / Tuesday. I will pop over if I have time to discuss this with you